### ISSUE TITLE
[QST] cuDF dependencies & UnsatisfiedLinkError

### ISSUE THREAD

**Issue by martinstuder** [NVIDIA/spark-rapids#1067]
**What is your question?**

I'm running into the following `UnsatisfiedLinkError` and I'm wondering what dependencies (or configurations) I'm missing:
```
Caused by: java.lang.UnsatisfiedLinkError: ai.rapids.cudf.ColumnVector.makeEmptyCudfColumn(I)J
	at ai.rapids.cudf.ColumnVector.makeEmptyCudfColumn(Native Method)
	at ai.rapids.cudf.ColumnVector.access$900(ColumnVector.java:40)
	at ai.rapids.cudf.ColumnVector$OffHeapState.<init>(ColumnVector.java:3060)
	at ai.rapids.cudf.ColumnVector.<init>(ColumnVector.java:171)
	at ai.rapids.cudf.HostColumnVector.copyToDevice(HostColumnVector.java:396)
	at ai.rapids.cudf.HostColumnVector$Builder.buildAndPutOnDevice(HostColumnVector.java:2040)
	at com.nvidia.spark.rapids.GpuColumnVector$GpuColumnarBatchBuilder.build(GpuColumnVector.java:104)
	at com.nvidia.spark.rapids.GpuColumnVector.emptyBatch(GpuColumnVector.java:217)
	at com.nvidia.spark.rapids.GpuColumnVector.emptyBatch(GpuColumnVector.java:232)
	at org.apache.spark.sql.rapids.execution.SerializeConcatHostBuffersDeserializeBatch.writeObject(GpuBroadcastExchangeExec.scala:72)
	at sun.reflect.GeneratedMethodAccessor480.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1154)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1496)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at com.esotericsoftware.kryo.serializers.JavaSerializer.write(JavaSerializer.java:51)
	at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:651)
	at org.apache.spark.serializer.KryoSerializationStream.writeObject(KryoSerializer.scala:259)
	at org.apache.spark.broadcast.TorrentBroadcast$.$anonfun$blockifyObject$4(TorrentBroadcast.scala:334)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1559)
	at org.apache.spark.broadcast.TorrentBroadcast$.blockifyObject(TorrentBroadcast.scala:336)
	at org.apache.spark.broadcast.TorrentBroadcast.writeBlocks(TorrentBroadcast.scala:154)
	at org.apache.spark.broadcast.TorrentBroadcast.<init>(TorrentBroadcast.scala:105)
	at org.apache.spark.broadcast.TorrentBroadcastFactory.newBroadcast(TorrentBroadcastFactory.scala:37)
	at org.apache.spark.broadcast.BroadcastManager.newBroadcast(BroadcastManager.scala:95)
	at org.apache.spark.SparkContext.broadcast(SparkContext.scala:1589)
	at org.apache.spark.sql.rapids.execution.GpuBroadcastExchangeExecBase$$anon$1.$anonfun$call$1(GpuBroadcastExchangeExec.scala:325)
```

Running with `LD_DEBUG=libs` I find:
```
594:	/usr/lib/libhadoop.so: error: symbol lookup error: undefined symbol: Java_ai_rapids_cudf_ColumnVector_makeEmptyCudfColumn (fatal)
       594:	/local_disk0/tmp/libzstd-jni9001325620576199886.so: error: symbol lookup error: undefined symbol: Java_ai_rapids_cudf_ColumnVector_makeEmptyCudfColumn (fatal)
       594:	/usr/lib/libhadoop.so: error: symbol lookup error: undefined symbol: Java_ai_rapids_cudf_ColumnVector_makeEmptyCudfColumn__I (fatal)
       594:	/local_disk0/tmp/libzstd-jni9001325620576199886.so: error: symbol lookup error: undefined symbol: Java_ai_rapids_cudf_ColumnVector_makeEmptyCudfColumn__I (fatal)
       594:	/usr/lib/libhadoop.so: error: symbol lookup error: undefined symbol: Java_ai_rapids_cudf_ColumnVector_makeEmptyCudfColumn (fatal)
       594:	/local_disk0/tmp/libzstd-jni9001325620576199886.so: error: symbol lookup error: undefined symbol: Java_ai_rapids_cudf_ColumnVector_makeEmptyCudfColumn (fatal)
       594:	/usr/lib/libhadoop.so: error: symbol lookup error: undefined symbol: Java_ai_rapids_cudf_ColumnVector_makeEmptyCudfColumn__I (fatal)
       594:	/local_disk0/tmp/libzstd-jni9001325620576199886.so: error: symbol lookup error: undefined symbol: Java_ai_rapids_cudf_ColumnVector_makeEmptyCudfColumn__I (fatal)
```
On the Spark environment tab I get the following `java.library.path`:
`/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/usr/java/packages/lib/amd64:/usr/lib/x86_64-linux-gnu/jni:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/usr/lib/jni:/lib:/usr/lib`
which should correctly point to the cuda libraries according to https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/10.2/ubuntu18.04-x86_64/base/Dockerfile

I also installed `libboost-filesystem` according to information found on https://github.com/rapidsai/cudf/tree/branch-0.15/java

**Environment details**
* Azure Databricks 7.0 with Databricks Container Services base image based on `nvidia/cuda:10.2-cudnn7-runtime-ubuntu18.04`
* rapids-4-spark v0.2.0 with cudf-0.15-cuda10-2 (CUDA 10.2)

**Comment by jlowe** [comment]
From the stacktrace it looks like this may be related to #975.  Do you have CUDA and a GPU on the Spark driver node?  If not then that's probably the issue.  The workaround for RAPIDS Accelerator 0.2 is to make sure the driver node setup looks just like the worker nodes regarding CUDA and GPU availability.  This has been fixed for version 0.3 if you are comfortable building the latest on branch-0.3 and giving that a shot (note branch-0.3 currently requires cudf-0.17-SNAPSHOT).

**Comment by martinstuder** [comment]
@jlowe Right, using a GPU-enabled driver node solved the problem. Thanks for the feedback!
